### PR TITLE
[WIP] Cfunc x86 abi

### PR DESCRIPTION
**Pull request by SabirAkhadov**
This PR makes CFunc produce LLVM IR function wrapper compliant with amd64 ABI specification and 
have the same function signature as expected by a clang compiler for a given C type. Most importantly it covers the struct handling. Most of the test cases were taken from https://github.com/scrossuk/llvm-abi.
Additionally, two new types are added, CVoid and Vector. CVoid is needed for the correct void return type. Vector type is expected by clang when two floats occur inside a struct. The VectorType class should eventually be added to llvmlite, but for the sake of making it pass the CI is included directly here.

### CODE REVIEW

**Comment by sklam** [comment]
Impressive!!!!

**Comment by sklam** [comment]
Do you think the arg layout logic can be refactored into something more standalone and then `CFunc` can just use that?

**Comment by sklam** [comment]
FYI, we have a callconv file at:   https://github.com/numba/numba/blob/master/numba/targets/callconv.py

**Comment by SabirAkhadov** [comment]
Yes, I think I could modify callconv.CPUCallConv.call_function directly to generate the right instructions.
I was not sure if this class is used by something else. 

With arg layout logic do you mean the _compute_arguments() method? I am not sure where to put that otherwise. 

Completely unrelated: there is this library which facilitates correct llvm-abi handling https://github.com/scrossuk/llvm-abi, I could not find a way to integrate it into numba, could it be possible to use it in llvmlite?

**Comment by sklam** [comment]
@SabirAkhadov, are you done with this PR feature-wise?

File: numba/datamodel/models.py [hunk 1]
```diff
@@ -370,6 +370,53 @@ def load_from_data_pointer(self, builder, ptr, align=None):
...
+
+    def __init__(self, dmm, fe_type):
+        dtype_model = dmm.lookup(fe_type.dtype)
+        count = fe_type.count
+        be_type = VectorType(count, dtype_model.get_value_type())
+        super(VectorModel, self).__init__(dmm, fe_type, be_type)
+
+
+#  XXX: this should be part of llvmlite
+class VectorType(ir.Type):
```
Comment (sklam) [inline numba/datamodel/models.py]: There are now 2 PR about VectorType in llvmlite.  Numba team needs to resolve them: 
* https://github.com/numba/llvmlite/pull/313
* https://github.com/numba/llvmlite/pull/322

**Comment by sklam** [comment]
I've kicked a build on the buildfarm.  Need to check back later.

**Comment by sklam** [comment]
The tests are not checking whether proper C-ABI is generated.  The tests should try to execute the cfunc using one of the `.ctypes` or `.cffi` wrappers.

**Comment by sklam** [comment]
Also, it seems like the amd64 ABI is used regardless of the host architecture.  Does this work on x86?  What about non-x86 platforms?  i.e ARM and Power?

**Comment by saulshanabrook** [comment]
What about using clang to generate this LLVM code? I am looking into using dragonffi to try to generate these wrappers: https://github.com/aguinet/dragonffi/issues/13

**Pull request closed without merging**